digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/RoadLinkService.scala

```diff
@@ -1143,40 +1142,42 @@ class RoadLinkService(val vvhClient: VVHClient, val eventbus: DigiroadEventBus,
-            }
-          } else {
-            val roadLinkAttributesRelatedWithThisChange = LinkAttributesDao.getExistingValues(change.oldId.get, Some(change.vvhTimeStamp))
+        ChangeType.apply(change.changeType) match {
+          case ChangeType.New =>
+            resolveNewChange(change)
+          case ChangeType.Removed =>
+            if (LinkAttributesDao.getExistingValues(change.oldId.get, Some(change.vvhTimeStamp)).nonEmpty)
+              LinkAttributesDao.expireValues(change.oldId.get, Some(changeUsername), Some(change.vvhTimeStamp))
+          case _ if LinkAttributesDao.getExistingValues(change.newId.get).isEmpty =>
+            val newIdFromVariousOld = changesToBeProcessed.filter(cp => cp.newId == change.newId)
+            if (newIdFromVariousOld.size > 1) {
+              val oldIdsAttributes = newIdFromVariousOld.map { thisChange =>
+                val attributes = LinkAttributesDao.getExistingValues(thisChange.oldId.get, Some(change.vvhTimeStamp))
+                attributes
+              }
 
-            if (roadLinkAttributesRelatedWithThisChange.nonEmpty) {
-              val newIdsOfThisOldId = changesToBeProcessed.filter(thisChange => thisChange.oldId == change.oldId && thisChange.newId.isDefined)
+              returnEqualAttributes(oldIdsAttributes).foreach { case (attribute, value) =>
+                LinkAttributesDao.insertAttributeValueByChanges(change.newId.get, changeUsername, attribute, value, change.vvhTimeStamp)
+              }
+            } else {
+              val roadLinkAttributesRelatedWithThisChange = LinkAttributesDao.getExistingValues(change.oldId.get, Some(change.vvhTimeStamp))
 
-              LinkAttributesDao.expireValues(change.oldId.get, Some(changeUsername), Some(change.vvhTimeStamp))
-              newIdsOfThisOldId.foreach { changeNewId =>
+              if (roadLinkAttributesRelatedWithThisChange.nonEmpty) {
                 roadLinkAttributesRelatedWithThisChange.foreach { case (attribute, value) =>
-                  LinkAttributesDao.insertAttributeValueByChanges(changeNewId.newId.get, changeUsername, attribute, value, changeNewId.vvhTimeStamp)
+                  LinkAttributesDao.insertAttributeValueByChanges(change.newId.get, changeUsername, attribute, value, change.vvhTimeStamp)
+
                 }
               }
             }
-          }
+          case _ =>
         }
       }
     }
 
-    val changesToBeProcessed = changes.filter(change => change.oldId != change.newId).sortWith(_.vvhTimeStamp < _.vvhTimeStamp)
+    val changesToBeProcessed = changes.filterNot(change => change.oldId == change.newId)
+      .filterNot(change => change.oldId.isDefined && change.newId.isEmpty && change.changeType != ChangeType.Removed.value)
+      .filterNot(change => change.oldId.isEmpty && change.newId.isDefined && change.changeType != ChangeType.New.value)
+      .sortWith(_.vvhTimeStamp < _.vvhTimeStamp)
```

Comment:
Sugestion:
`hanges.filterNot{ change => 
	val isOldEqNew = change.oldId == change.newId
	val newIdAndNotToRemoved = change.oldId.isDefined && change.newId.isEmpty && change.changeType != ChangeType.Removed.value
	val NewAndNotToNew = change.oldId.isEmpty && change.newId.isDefined && change.changeType != ChangeType.New.value
	
	isOldEqNew || newIdAndNotToRemoved || NewAndNotToNew

}.sortWith(_.vvhTimeStamp < _.vvhTimeStamp)`

Reply:
I agree with you except in the vals

```
val changesToBeProcessed = changes.filterNot(change => change.oldId == change.newId || 
(change.oldId.isDefined && change.newId.isEmpty && change.changeType != ChangeType.Removed.value) || 
(change.oldId.isEmpty && change.newId.isDefined && change.changeType != ChangeType.New.value))
      .sortWith(_.vvhTimeStamp < _.vvhTimeStamp)
```

Reply:
For me, I prefer the @ricardofmsimoes approach but don't like the names of the variables in particular "NewAndNotToNew" variable starting with big letters.

Reply:
val isOldEqNew = change.oldId == change.newId
      val isNotToExpire = change.oldId.isDefined && change.newId.isEmpty && change.changeType != ChangeType.Removed.value
      val isNotToCreate = change.oldId.isEmpty && change.newId.isDefined && change.changeType != ChangeType.New.value

Names applied to the vals